Atlas dynamics demo

amcastro-tri opened the pull request:
Addresses reviews on original PR #2164 by @RussTedrake which implements a dynamics-only simulation of Atlas.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/2542)
<!-- Reviewable:end -->

amcastro-tri commented:
+@david-german-tri for platform review
+@sherm1 and +@russtedrake for feature review

<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> #### Atlas dynamics demo
> 
> Addresses reviews on original PR #2164 by @RussTedrake which implements a dynamics-only simulation of Atlas.
> 
> 
> 
</details>


---

Review status: 0 of 5 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2542#-:-KJvkt7g0iB8zwRnZxbj:-1648508500)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
cc'ing @tkoolen interested in Valkyrie.

RussTedrake commented:
Reviewed 2 of 6 files at r2, 3 of 3 files at r3.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/examples/Atlas/atlas_system.cc, line 9 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2542#-KK5TsiUyLck2lkyMbtx:-KK5TsiUyLck2lkyMbty:252247791) ([raw file](https://github.com/robotlocomotion/drake/blob/392953241f76b5370f5311a9aae285efde998275/drake/examples/Atlas/atlas_system.cc#L9)):*
> ```cpp
> using Eigen::Isometry3d;
> 
> AtlasSystem::AtlasSystem() {
> ```

i don't like the name.  :)  in matlab, we call this class simply Atlas.  AtlasPlant might be a reasonable name, too.  

---

*[drake/examples/Atlas/atlas_system.h, line 18 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2542#-KK5UEYlBaCKJkaCyCCJ:-KK5UEYlBaCKJkaCyCCK:-1293033753) ([raw file](https://github.com/robotlocomotion/drake/blob/392953241f76b5370f5311a9aae285efde998275/drake/examples/Atlas/atlas_system.h#L18)):*
> ```objc
> 
>  private:
>   RigidBodyTree* tree_;
> ```

why keep a pointer to the tree, when you already have one via the rigidbodysystem?  bad things could happen if they get out of sync.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2542)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Reviewed changes. I didn't try executing this -- what happens when it runs, @amcastro-tri?

<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> cc'ing @tkoolen interested in Valkyrie.
</details>


---

Reviewed 2 of 6 files at r2, 3 of 3 files at r3.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/examples/Atlas/atlas_system.cc, line 9 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2542#-KK5TsiUyLck2lkyMbtx:-KK79d9wZc5_7HX0lamJ:-177860007) ([raw file](https://github.com/robotlocomotion/drake/blob/392953241f76b5370f5311a9aae285efde998275/drake/examples/Atlas/atlas_system.cc#L9)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary>
> i don't like the name.  :)  in matlab, we call this class simply Atlas.  AtlasPlant might be a reasonable name, too.  
</details>
Why don't you like that, Russ? It is a kind of RigidBodySystem so satisfies the System concept, and it does have Atlas built in -- I would have thought this is a good name.

---

*[drake/examples/Atlas/atlas_system.h, line 11 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2542#-KK7AK-IWp4owcapVIvu:-KK7AK-JvcuusS8kiq6-:2087875128) ([raw file](https://github.com/robotlocomotion/drake/blob/392953241f76b5370f5311a9aae285efde998275/drake/examples/Atlas/atlas_system.h#L11)):*
> ```objc
> namespace drake {
> 
> class AtlasSystem : public Drake::RigidBodySystem {
> ```

Please add some class documentation.

---

*[drake/examples/Atlas/atlas_system.h, line 23 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2542#-KK7AflTFmBlmn6wdXL5:-KK7AflTFmBlmn6wdXL6:-1734088099) ([raw file](https://github.com/robotlocomotion/drake/blob/392953241f76b5370f5311a9aae285efde998275/drake/examples/Atlas/atlas_system.h#L23)):*
> ```objc
>   // Sets the initial pose for Atlas.
>   // Magic numbers are initial conditions used in runAtlasWalking.m.
>   void SetInitialConfiguration();
> ```

There aren't any magic numbers here -- that's in the implementation so the comment should go there.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2542#-:-KK7BsJwyikGUdKKELEV:658525154)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
The simulation starts with Atlas in a standing position with the knees slightly bent. There are no controllers. Atlas falls to the ground like a rag doll. There is a clear penetration into the ground due to our spring model of contact. Now that you ask, I will try to play around with these and see if I can increase the time step a bit.

<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> Reviewed changes. I didn't try executing this -- what happens when it runs, @amcastro-tri?
</details>


---

Review status: 3 of 5 files reviewed at latest revision, 5 unresolved discussions.

---

*[drake/examples/Atlas/atlas_system.cc, line 9 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2542#-KK5TsiUyLck2lkyMbtx:-KK8aoWgWEFUD1vn_lhN:-1199335683) ([raw file](https://github.com/robotlocomotion/drake/blob/392953241f76b5370f5311a9aae285efde998275/drake/examples/Atlas/atlas_system.cc#L9)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> Why don't you like that, Russ? It is a kind of RigidBodySystem so satisfies the System concept, and it does have Atlas built in -- I would have thought this is a good name.
</details>
Like @sherm1 explained, that is exactly why I chose that name. However, more than happy to change the name if we agree on a convention or we'd like to follow a convention already established in Matlab (I am sorry but I do not know what the convention in Matlab is if any). With System 2.0 promoting inheritance we will have lots of Systems inheriting from either SystemInterface of DiagramInterface. No doubt a good idea to start agreeing on what to call such systems.

---

*[drake/examples/Atlas/atlas_system.h, line 11 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2542#-KK7AK-IWp4owcapVIvu:-KK8jm7cTiA1zHSzHCku:-1063518331) ([raw file](https://github.com/robotlocomotion/drake/blob/392953241f76b5370f5311a9aae285efde998275/drake/examples/Atlas/atlas_system.h#L11)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> Please add some class documentation.
</details>
Added doc. 

---

*[drake/examples/Atlas/atlas_system.h, line 18 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2542#-KK5UEYlBaCKJkaCyCCJ:-KK8cJN7m0cE-4HW_uvI:1812567124) ([raw file](https://github.com/robotlocomotion/drake/blob/392953241f76b5370f5311a9aae285efde998275/drake/examples/Atlas/atlas_system.h#L18)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary>
> why keep a pointer to the tree, when you already have one via the rigidbodysystem?  bad things could happen if they get out of sync.
</details>
`RigidBodySystem::tree` is private. Therefore I defined this variable as a convenience so that I would not need to type the lengthier variant `getRigidBodyTree()` :-). It could not get out of sync since it is a pointer to the original RigidBodyTree in RigidBodySystem and initialized in the constructor.
However, I can see how this style could bother you. I removed `AltalsSystem::tree_` and decided to define a convenience `tree` locally where needed. PTAL

---

*[drake/examples/Atlas/atlas_system.h, line 23 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2542#-KK7AflTFmBlmn6wdXL5:-KK8eOuw16Ew1pIvbrca:-164386170) ([raw file](https://github.com/robotlocomotion/drake/blob/392953241f76b5370f5311a9aae285efde998275/drake/examples/Atlas/atlas_system.h#L23)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> There aren't any magic numbers here -- that's in the implementation so the comment should go there.
</details>
Agree. Moved comment to implementation where it belongs.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2542#-:-KK8a-Z2JFnOJUR9EU5E:-313308094)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: 3 of 5 files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/examples/Atlas/atlas_system.cc, line 9 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2542#-KK5TsiUyLck2lkyMbtx:-KK8yC_5_zMaC3oTLIpR:-1602757438) ([raw file](https://github.com/robotlocomotion/drake/blob/392953241f76b5370f5311a9aae285efde998275/drake/examples/Atlas/atlas_system.cc#L9)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> Like @sherm1 explained, that is exactly why I chose that name. However, more than happy to change the name if we agree on a convention or we'd like to follow a convention already established in Matlab (I am sorry but I do not know what the convention in Matlab is if any). With System 2.0 promoting inheritance we will have lots of Systems inheriting from either SystemInterface of DiagramInterface. No doubt a good idea to start agreeing on what to call such systems.
</details>
Actually, we already have a System 2.0 example of this with the [SpringMassSystem example](https://github.com/RobotLocomotion/drake/blob/master/drake/examples/spring_mass/spring_mass_system.h#L71).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2542)*
<!-- Sent from Reviewable.io -->

david-german-tri commented:
Review status: 3 of 5 files reviewed at latest revision, 5 unresolved discussions.

---

*[drake/examples/Atlas/atlas_system.cc, line 9 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2542#-KK5TsiUyLck2lkyMbtx:-KK9r7qEklc4d43OYKyN:-1090216026) ([raw file](https://github.com/robotlocomotion/drake/blob/392953241f76b5370f5311a9aae285efde998275/drake/examples/Atlas/atlas_system.cc#L9)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> Actually, we already have a System 2.0 example of this with the [SpringMassSystem example](https://github.com/RobotLocomotion/drake/blob/master/drake/examples/spring_mass/spring_mass_system.h#L71).
</details>
I don't feel strongly about it, but I do think `FooSystem` is a helpful convention.

---

*[drake/examples/Atlas/atlas_system.h, line 19 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2542#-KK9sFqy-_ekfbi1glbf:-KK9sFqzG6X0ZcO1v-d-:1606243797) ([raw file](https://github.com/robotlocomotion/drake/blob/146bce7e55526d6266b41073446611077fe5cc45/drake/examples/Atlas/atlas_system.h#L19)):*
> ```objc
>  sensors, controllers and, methods implementing specific capabilities for Atlas.
>  **/
> class AtlasSystem : public Drake::RigidBodySystem {
> ```

#2550->

This seems like a classic case where composition is better than inheritance.  It doesn't make sense for `AtlasSystem` to provide methods like `addRobotFromURDF`.  It should be pretty easy to satisfy the System concept by implementing a few forwarding methods.

https://google.github.io/styleguide/cppguide.html#Inheritance

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2542)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: 3 of 5 files reviewed at latest revision, 5 unresolved discussions.

---

*[drake/examples/Atlas/atlas_system.h, line 19 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2542#-KK9sFqy-_ekfbi1glbf:-KKA7MYS4soerWZWkrz_:-1362681992) ([raw file](https://github.com/robotlocomotion/drake/blob/146bce7e55526d6266b41073446611077fe5cc45/drake/examples/Atlas/atlas_system.h#L19)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> #2550->
> 
> This seems like a classic case where composition is better than inheritance.  It doesn't make sense for `AtlasSystem` to provide methods like `addRobotFromURDF`.  It should be pretty easy to satisfy the System concept by implementing a few forwarding methods.
> 
> https://google.github.io/styleguide/cppguide.html#Inheritance
> 
</details>
To me an `AtlasSystem` "is a" `RigidBodySystem`, and therefore inheritance. IMO a method like `addRobotFromURDF` does not belong there but to another "Loader", but that's another issue.  

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2542)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: 3 of 5 files reviewed at latest revision, 5 unresolved discussions.

---

*[drake/examples/Atlas/atlas_system.h, line 19 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2542#-KK9sFqy-_ekfbi1glbf:-KKABe8cQ3VX6pqDlSBf:-1461761384) ([raw file](https://github.com/robotlocomotion/drake/blob/146bce7e55526d6266b41073446611077fe5cc45/drake/examples/Atlas/atlas_system.h#L19)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> To me an `AtlasSystem` "is a" `RigidBodySystem`, and therefore inheritance. IMO a method like `addRobotFromURDF` does not belong there but to another "Loader", but that's another issue.  
</details>
I see the current confusion as stemming from the fact that there's only _one_ `RigidBodySystem`.  Maybe once Drake is modified to support _multiple_ `RigidBodySystem` objects all within a `World`, an inheritance model will make more sense.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2542)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: 3 of 5 files reviewed at latest revision, 5 unresolved discussions.

---

*[drake/examples/Atlas/atlas_system.h, line 19 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2542#-KK9sFqy-_ekfbi1glbf:-KKAClnX0a3yDneEO4Aa:-1912894091) ([raw file](https://github.com/robotlocomotion/drake/blob/146bce7e55526d6266b41073446611077fe5cc45/drake/examples/Atlas/atlas_system.h#L19)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> I see the current confusion as stemming from the fact that there's only _one_ `RigidBodySystem`.  Maybe once Drake is modified to support _multiple_ `RigidBodySystem` objects all within a `World`, an inheritance model will make more sense.
</details>
@liangfok: that is a really good example. Thanks

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2542)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: 3 of 5 files reviewed at latest revision, 5 unresolved discussions.

---

*[drake/examples/Atlas/atlas_system.h, line 19 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2542#-KK9sFqy-_ekfbi1glbf:-KKDduBwT0bFUX4v-qST:887674037) ([raw file](https://github.com/robotlocomotion/drake/blob/146bce7e55526d6266b41073446611077fe5cc45/drake/examples/Atlas/atlas_system.h#L19)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> @liangfok: that is a really good example. Thanks
</details>
As an interesting fact, Kitware's VTK implements this idea of ["Readers"](http://www.vtk.org/Wiki/VTK/Examples/Cxx/IO/ReadOBJ) as separate concepts. I like that idea. 
@david-german-tri: Probably a f2f will help me understand better your preference towards compositions? My intention here was to replicate the inheritance concept between `ContinuousSystem` and `SpringMassSystem`. Right now probably our RigidBodySystem 1.0 design does not fully support this scheme, but I like to start thinking on how we will be doing things with System 2.0. And I though this scheme of inheriting from a more general system was one of the advantages of System 2.0

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2542)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: 3 of 5 files reviewed at latest revision, 5 unresolved discussions.

---

*[drake/examples/Atlas/atlas_system.h, line 19 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2542#-KK9sFqy-_ekfbi1glbf:-KKDiiu3isxdwplwxGJE:1014898001) ([raw file](https://github.com/robotlocomotion/drake/blob/146bce7e55526d6266b41073446611077fe5cc45/drake/examples/Atlas/atlas_system.h#L19)):*
> As an interesting fact, Kitware's VTK implements this idea of "Readers" as separate concepts.

Cool! Should we employ a similar design for reading URDF / SDF / XODR / RoadXML files?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2542)*
<!-- Sent from Reviewable.io -->

david-german-tri commented:
Review status: 3 of 5 files reviewed at latest revision, 5 unresolved discussions.

---

*[drake/examples/Atlas/atlas_system.h, line 19 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2542#-KK9sFqy-_ekfbi1glbf:-KKEAclNwEpEY0vt2vqd:1805140491) ([raw file](https://github.com/robotlocomotion/drake/blob/146bce7e55526d6266b41073446611077fe5cc45/drake/examples/Atlas/atlas_system.h#L19)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> > As an interesting fact, Kitware's VTK implements this idea of "Readers" as separate concepts.
> 
> Cool! Should we employ a similar design for reading URDF / SDF / XODR / RoadXML files?
</details>
First off, I agree it would be good to get the parse methods out of RigidBodySystem altogether.  This PR probably isn't the right place to choose an approach.  I mentioned the parse method only as a handy example of how inheritance can over-share.  Rule of thumb: inheritance is appropriate when you need to share interface, whereas composition is appropriate when you only need to share implementation.  

`ContinuousSystem` is definitely an interface contract point.  My feeling is that `RigidBodySystem` is not.  It's not clear-cut, though; I could imagine you persuading me otherwise.  So, yes, happy to talk f2f.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2542)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: 3 of 5 files reviewed at latest revision, 5 unresolved discussions.

---

*[drake/examples/Atlas/atlas_system.h, line 19 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2542#-KK9sFqy-_ekfbi1glbf:-KKEVK1nPzmyJwmjTDTq:-716457633) ([raw file](https://github.com/robotlocomotion/drake/blob/146bce7e55526d6266b41073446611077fe5cc45/drake/examples/Atlas/atlas_system.h#L19)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> First off, I agree it would be good to get the parse methods out of RigidBodySystem altogether.  This PR probably isn't the right place to choose an approach.  I mentioned the parse method only as a handy example of how inheritance can over-share.  Rule of thumb: inheritance is appropriate when you need to share interface, whereas composition is appropriate when you only need to share implementation.  
> 
> `ContinuousSystem` is definitely an interface contract point.  My feeling is that `RigidBodySystem` is not.  It's not clear-cut, though; I could imagine you persuading me otherwise.  So, yes, happy to talk f2f.
</details>
Actually I realized this morning that probably in the future we'd like to update `AtlasSystem` to inherit from `Diagram`. If so, then `AtlasSystem` would inherit the system's interface while at the same time it'd be composed of several systems (since Diagram is). I think that'd be the right thing; `AtlasSystem` would be a system composed of several other systems like `RigidBodySystem` a `WalkingControllerSystem`, a number of `HydarulicActuatorSystem`, etc. all connected (compose) by Diagram. 
I can see however how System 1.0 does not support this idea. 
And I do agree that the side conversation on parsers is out of the scope of this PR, but a fun discussion though :-)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2542)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
This PR generated some useful discussion. It is not clear to me if such comments are blocking this PR or not. I'd appreciate if you can take a look whenever you have a chance.

---

Review status: 3 of 5 files reviewed at latest revision, 5 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2542#-:-KKTbLyHuYtj2Zja2-t7:-1134060958)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Just one BTW comment from me (see below).

---

Reviewed 2 of 2 files at r4.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[drake/examples/Atlas/atlas_system.h, line 19 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2542#-KK9sFqy-_ekfbi1glbf:-KKTicTixZpKdXKleCbH:1399211241) ([raw file](https://github.com/robotlocomotion/drake/blob/146bce7e55526d6266b41073446611077fe5cc45/drake/examples/Atlas/atlas_system.h#L19)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> Actually I realized this morning that probably in the future we'd like to update `AtlasSystem` to inherit from `Diagram`. If so, then `AtlasSystem` would inherit the system's interface while at the same time it'd be composed of several systems (since Diagram is). I think that'd be the right thing; `AtlasSystem` would be a system composed of several other systems like `RigidBodySystem` a `WalkingControllerSystem`, a number of `HydarulicActuatorSystem`, etc. all connected (compose) by Diagram. 
> I can see however how System 1.0 does not support this idea. 
> And I do agree that the side conversation on parsers is out of the scope of this PR, but a fun discussion though :-)
</details>
BTW I believe we decided against inheritance and instead want to use composition here. No?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2542#-:-KKTiiuUaRl2_4DWrVaf:790915832)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/examples/Atlas/atlas_system.cc, line 9 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2542#-KK5TsiUyLck2lkyMbtx:-KKUyBEpLPpf_ISwq-I9:-1244804804) ([raw file](https://github.com/robotlocomotion/drake/blob/392953241f76b5370f5311a9aae285efde998275/drake/examples/Atlas/atlas_system.cc#L9)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> I don't feel strongly about it, but I do think `FooSystem` is a helpful convention.
</details>
Let's leave this name as-is in this PR so it won't get blocked and table this topic for now. I'm unblocking this thread.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2542#-:-KKUztdKeqTuWbee_yYm:1417088721)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Looks there are a few blocked reviewer threads that need acknowledgment: @david-german-tri and @russtedrake each have one.

---

Reviewed 2 of 2 files at r4.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2542#-:-KKV-aKHZr_MbOkcQkhQ:-1285090784)*
<!-- Sent from Reviewable.io -->

jamiesnape commented:
@drake-jenkins-bot retest this please

jamiesnape commented:
@drake-jenkins-bot retest this please

RussTedrake commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/examples/Atlas/atlas_system.cc, line 9 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2542#-KK5TsiUyLck2lkyMbtx:-KLzlj_dfvvl-N3MaHn4:1747995772) ([raw file](https://github.com/robotlocomotion/drake/blob/392953241f76b5370f5311a9aae285efde998275/drake/examples/Atlas/atlas_system.cc#L9)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> Let's leave this name as-is in this PR so it won't get blocked and table this topic for now. I'm unblocking this thread.
</details>
Running Atlas in real life invokes about 20 different processes each doing complex system tasks.  In my mind, that stack is the Atlas System.  This is just one sliver -- the plant model.  I would prefer AtlasPlant or Atlas. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2542)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
As requested I used composition instead of inheritance. I also renamed `AtlasSystem` to be an `AtlasPlant`. `AtlasPlant` implements the system concept. PTAL.

---

Review status: 1 of 5 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/examples/Atlas/atlas_plant.cc, line 9 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2542#-KK5TsiUyLck2lkyMbtx:-KM5AClH_g7Wn8aVzzp9:426409381) ([raw file](https://github.com/robotlocomotion/drake/blob/392953241f76b5370f5311a9aae285efde998275/drake/examples/Atlas/atlas_plant.cc#L9)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary>
> Running Atlas in real life invokes about 20 different processes each doing complex system tasks.  In my mind, that stack is the Atlas System.  This is just one sliver -- the plant model.  I would prefer AtlasPlant or Atlas. 
</details>
I like that. I renamed `AtlasSystem` to be `AtlasPlant`.

---

*[drake/examples/Atlas/atlas_system.h, line 19 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2542#-KK9sFqy-_ekfbi1glbf:-KM5AKg1D-iRzKWMdWkb:-447020726) ([raw file](https://github.com/robotlocomotion/drake/blob/146bce7e55526d6266b41073446611077fe5cc45/drake/examples/Atlas/atlas_system.h#L19)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> BTW I believe we decided against inheritance and instead want to use composition here. No?
</details>
System 1.0 concepts are not very clear (at least to me). Inheritance or composition are just an implementation detail at this stage. Eager to get System 2.0 online. 
Now I decided to use composition if that's what makes everybody happy.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2542#-:-KM59nCOvv3MiGYLxUWe:-870951736)*
<!-- Sent from Reviewable.io -->

jamiesnape commented:
@drake-jenkins-bot retest this please

sherm1 commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> again.

---

Review status: 1 of 5 files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2542#-:-KM5GJugxegjERUD3M9Z:1863939325)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Done with another round of review.

---

Reviewed 1 of 7 files at r2, 6 of 6 files at r6.
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---

*[drake/examples/Atlas/atlas_plant.h, line 11 \[r6\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2542#-KMB-EQLlrQ0GmvMn40U:-KMB-EQMIJ7cj2Mnq7dX:-292154172) ([raw file](https://github.com/robotlocomotion/drake/blob/851de19e41a895c701b202e6580e35a6795d32c0/drake/examples/Atlas/atlas_plant.h#L11)):*
> ```objc
> namespace drake {
> 
> /** Implements the System concept for the Atlas robot.
> ```

BTW, I believe we're standardizing on using `/// foo...` as the standard Doxygen block notation.

(This comment applies to all Doxygen blocks in this PR.)
(See, for example, the code in System 2.0.)

---

*[drake/examples/Atlas/atlas_plant.h, line 13 \[r6\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2542#-KMAzppKbDsin7_BRr1L:-KMAzppKbDsin7_BRr1M:-860288845) ([raw file](https://github.com/robotlocomotion/drake/blob/851de19e41a895c701b202e6580e35a6795d32c0/drake/examples/Atlas/atlas_plant.h#L13)):*
> ```objc
> /** Implements the System concept for the Atlas robot.
> 
> This class is implemented to conveniently place all methods and functionality
> ```

BTW, consider changing "place" to "hold" and "functionality" to "state".

---

*[drake/examples/Atlas/atlas_plant.h, line 46 \[r6\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2542#-KMB-fUNRPq33ykaphaX:-KMB-fUNRPq33ykaphaY:-1760213933) ([raw file](https://github.com/robotlocomotion/drake/blob/851de19e41a895c701b202e6580e35a6795d32c0/drake/examples/Atlas/atlas_plant.h#L46)):*
> ```objc
>                              const InputVector<double>& u) const;
> 
>   /** Given a an InputVector @p u of generalized forces this method computes the
> ```

typo "a an"

Also, I believe there should be a comma after "forces".

---

*[drake/examples/Atlas/atlas_plant.h, line 53 \[r6\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2542#-KMB-qbdYlFLtcjGXAvR:-KMB-qbeBk4JxG7iRtlS:-1799326653) ([raw file](https://github.com/robotlocomotion/drake/blob/851de19e41a895c701b202e6580e35a6795d32c0/drake/examples/Atlas/atlas_plant.h#L53)):*
> ```objc
> 
>  private:
>   // The underlying rigid body system
> ```

BTW, missing period at end of sentence.

---

*[drake/systems/plants/rigidBodyLCMNode.cpp, line 37 \[r6\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2542#-KMB03pReftT52-QE_5z:-KMB03pReftT52-QE_6-:555828939) ([raw file](https://github.com/robotlocomotion/drake/blob/851de19e41a895c701b202e6580e35a6795d32c0/drake/systems/plants/rigidBodyLCMNode.cpp#L37)):*
> ```cpp
>               << "Options: " << std::endl
>               << "  --base [FIXED|RPY|(QUAT)]      defines the connection "
>                  "between the root link and the world"
> ```

BTW, "root link" may be confusing. How about "Atlas robot"?

---

*[drake/systems/plants/rigidBodyLCMNode.cpp, line 40 \[r6\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2542#-KMB0B5sdqBI5l-QG2Cd:-KMB0B5tNZJITF8ayL1S:-1854491955) ([raw file](https://github.com/robotlocomotion/drake/blob/851de19e41a895c701b202e6580e35a6795d32c0/drake/systems/plants/rigidBodyLCMNode.cpp#L40)):*
> ```cpp
>               << std::endl
>               << "  --add_flat_terrain             adds a large box to "
>                  "approximate flat terrain"
> ```

BTW, missing "a" after "approximate".

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2542#-:-KMB5jPBce4ygjy5xq7U:-403560003)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/examples/Atlas/atlas_plant.h, line 11 \[r6\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2542#-KMB-EQLlrQ0GmvMn40U:-KMBAB62kzrhV_DHjiXi:-1138777860) ([raw file](https://github.com/robotlocomotion/drake/blob/851de19e41a895c701b202e6580e35a6795d32c0/drake/examples/Atlas/atlas_plant.h#L11)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> BTW, I believe we're standardizing on using `/// foo...` as the standard Doxygen block notation.
> 
> (This comment applies to all Doxygen blocks in this PR.)
> (See, for example, the code in System 2.0.)
</details>
FYI, I do not agree with that at all @liangfok! I hate all that unnecessary punctuation adding ugliness and eating up precious space, not to mention the difficulty of formatting the comments. I much prefer the clean way @amcastro-tri is doing it here and I actually don't understand why anyone would prefer the ugly approach. In any case we have definitely *not* standardized on it!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2542)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/examples/Atlas/atlas_plant.h, line 11 \[r6\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2542#-KMB-EQLlrQ0GmvMn40U:-KMBF406Lv479w6dU2it:1890246630) ([raw file](https://github.com/robotlocomotion/drake/blob/851de19e41a895c701b202e6580e35a6795d32c0/drake/examples/Atlas/atlas_plant.h#L11)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> FYI, I do not agree with that at all @liangfok! I hate all that unnecessary punctuation adding ugliness and eating up precious space, not to mention the difficulty of formatting the comments. I much prefer the clean way @amcastro-tri is doing it here and I actually don't understand why anyone would prefer the ugly approach. In any case we have definitely *not* standardized on it!
</details>
I was simply repeating a comment from @ggould-tri  here:

https://reviewable.io/reviews/robotlocomotion/drake/2712#-KM9XDFWFqyRzBZbn8in

Not sure if / how we will ever resolve #2051 without gentle nudges in a particular direction. :sweat_smile: 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2542)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/examples/Atlas/atlas_plant.h, line 11 \[r6\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2542#-KMB-EQLlrQ0GmvMn40U:-KMBNkgj-NIxfhyjo4ef:819578866) ([raw file](https://github.com/robotlocomotion/drake/blob/851de19e41a895c701b202e6580e35a6795d32c0/drake/examples/Atlas/atlas_plant.h#L11)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> I was simply repeating a comment from @ggould-tri  here:
> 
> https://reviewable.io/reviews/robotlocomotion/drake/2712#-KM9XDFWFqyRzBZbn8in
> 
> Not sure if / how we will ever resolve #2051 without gentle nudges in a particular direction. :sweat_smile: 
</details>
FYI, take a closer look at Grant's comment. He was saying to use a comment that would appear in doxygen, rather than the `//` comment that was in the PR and is invisible to doxygen. He was not (at least there) enforcing a particular doxygen style. [Dimitri himself](https://www.stack.nl/~dimitri/doxygen/manual/docblocks.html) lists the `/**` version first, due to its legacy from Javadoc, along with an absurd number of alternatives!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2542)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/examples/Atlas/atlas_plant.h, line 11 \[r6\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2542#-KMB-EQLlrQ0GmvMn40U:-KMByboMsImNTUW3C2xI:-551490578) ([raw file](https://github.com/robotlocomotion/drake/blob/851de19e41a895c701b202e6580e35a6795d32c0/drake/examples/Atlas/atlas_plant.h#L11)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> FYI, take a closer look at Grant's comment. He was saying to use a comment that would appear in doxygen, rather than the `//` comment that was in the PR and is invisible to doxygen. He was not (at least there) enforcing a particular doxygen style. [Dimitri himself](https://www.stack.nl/~dimitri/doxygen/manual/docblocks.html) lists the `/**` version first, due to its legacy from Javadoc, along with an absurd number of alternatives!
</details>
Yeah, I'm also more accustomed to the `/** ... **/`  and `@param` syntax because my "native" OO programming language is Java. Perhaps we should start standardizing on that syntax throughout Drake, including in System 2.0 components.

Oh well, marking this thread as "satisfied".

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2542)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/examples/Atlas/atlas_plant.h, line 11 \[r6\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2542#-KMB-EQLlrQ0GmvMn40U:-KMC0wew1Xkmok-Qz5yO:1326674726) ([raw file](https://github.com/robotlocomotion/drake/blob/851de19e41a895c701b202e6580e35a6795d32c0/drake/examples/Atlas/atlas_plant.h#L11)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> Yeah, I'm also more accustomed to the `/** ... **/`  and `@param` syntax because my "native" OO programming language is Java. Perhaps we should start standardizing on that syntax throughout Drake, including in System 2.0 components.
> 
> Oh well, marking this thread as "satisfied".
</details>
BTW, currently it is author's choice. I think everyone should use a style they find appealing and practical so others can see well-executed alternatives as the PRs go by. Authors who haven't used doxygen much yet could experiment to see what they like. Maybe one style will dominate on its merits, maybe not. These are just comments so it is hard to get too upset about the input format as long as it is readable and produces great API documentation.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2542)*
<!-- Sent from Reviewable.io -->

liangfok commented:
*[drake/examples/Atlas/atlas_plant.h, line 11 \[r6\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2542#-KMB-EQLlrQ0GmvMn40U:-KMCI-pgs2J9GH3CDy4y:-1213429558) ([raw file](https://github.com/robotlocomotion/drake/blob/851de19e41a895c701b202e6580e35a6795d32c0/drake/examples/Atlas/atlas_plant.h#L11)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> BTW, currently it is author's choice. I think everyone should use a style they find appealing and practical so others can see well-executed alternatives as the PRs go by. Authors who haven't used doxygen much yet could experiment to see what they like. Maybe one style will dominate on its merits, maybe not. These are just comments so it is hard to get too upset about the input format as long as it is readable and produces great API documentation.
</details>
SGTM :thumbsup: 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2542)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: 4 of 5 files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/examples/Atlas/atlas_plant.cc, line 9 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2542#-KK5TsiUyLck2lkyMbtx:-KMF9SlBxAFf2c8MvhJV:793965132) ([raw file](https://github.com/robotlocomotion/drake/blob/392953241f76b5370f5311a9aae285efde998275/drake/examples/Atlas/atlas_plant.cc#L9)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> I like that. I renamed `AtlasSystem` to be `AtlasPlant`.
</details>
@russtedrake: does this solution resolve this issue?

---

*[drake/examples/Atlas/atlas_plant.h, line 11 \[r6\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2542#-KMB-EQLlrQ0GmvMn40U:-KMFAHqpiqFelWQwiCGX:-613081746) ([raw file](https://github.com/robotlocomotion/drake/blob/851de19e41a895c701b202e6580e35a6795d32c0/drake/examples/Atlas/atlas_plant.h#L11)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> SGTM :thumbsup: 
</details>
We should just make the decision to use one form of documentation across all of Drake and stick with it. I personally do not like the `///` approach because as @sherm1 says it eats a lot of space and at for me sometimes I get confused what is public Doxygen and what is not specially is there are `//` comments intermixed with `///` documentation (I've seen it). With `/** **/` it is super clear what is Doxygen and what is just a comment for developers.
@liangfok, I just added this as a topic for our simulation meeting next Friday. I added a title for Friday 15 and a subtitle with this topic.

---

*[drake/examples/Atlas/atlas_plant.h, line 46 \[r6\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2542#-KMB-fUNRPq33ykaphaX:-KMF8sYojpv1-Or0Tkmk:-2137399250) ([raw file](https://github.com/robotlocomotion/drake/blob/851de19e41a895c701b202e6580e35a6795d32c0/drake/examples/Atlas/atlas_plant.h#L46)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> typo "a an"
> 
> Also, I believe there should be a comma after "forces".
</details>
Thanks

---

*[drake/examples/Atlas/atlas_system.h, line 19 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2542#-KK9sFqy-_ekfbi1glbf:-KMF9EREgpUK1khAbC4N:-1546367915) ([raw file](https://github.com/robotlocomotion/drake/blob/146bce7e55526d6266b41073446611077fe5cc45/drake/examples/Atlas/atlas_system.h#L19)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> System 1.0 concepts are not very clear (at least to me). Inheritance or composition are just an implementation detail at this stage. Eager to get System 2.0 online. 
> Now I decided to use composition if that's what makes everybody happy.
</details>
This comment status is still unresolved for @david-german-tri  

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2542)*
<!-- Sent from Reviewable.io -->

liangfok commented:
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> 

---

Reviewed 1 of 1 files at r7.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/examples/Atlas/atlas_plant.h, line 11 \[r6\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2542#-KMB-EQLlrQ0GmvMn40U:-KMFLlA03kAp2WSAOMEV:-447411688) ([raw file](https://github.com/robotlocomotion/drake/blob/851de19e41a895c701b202e6580e35a6795d32c0/drake/examples/Atlas/atlas_plant.h#L11)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> We should just make the decision to use one form of documentation across all of Drake and stick with it. I personally do not like the `///` approach because as @sherm1 says it eats a lot of space and at for me sometimes I get confused what is public Doxygen and what is not specially is there are `//` comments intermixed with `///` documentation (I've seen it). With `/** **/` it is super clear what is Doxygen and what is just a comment for developers.
> @liangfok, I just added this as a topic for our simulation meeting next Friday. I added a title for Friday 15 and a subtitle with this topic.
</details>
:thumbsup: Thanks for forcing the issue on our weekly group meeting. :smiley: 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2542#-:-KMFM6ugAzUXLbjsKths:-1868698381)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/examples/Atlas/atlas_system.h, line 19 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2542#-KK9sFqy-_ekfbi1glbf:-KMPLiesFdpO2Wlnny2-:11357076) ([raw file](https://github.com/robotlocomotion/drake/blob/146bce7e55526d6266b41073446611077fe5cc45/drake/examples/Atlas/atlas_system.h#L19)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> This comment status is still unresolved for @david-german-tri  
</details>
@david-german-tri: this is the last unresolved discussion for this PR. I changed the code to use composition instead. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2542)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Reviewed 5 of 6 files at r6, 1 of 1 files at r7.
Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2542)*
<!-- Sent from Reviewable.io -->
